### ISSUE TITLE
Mnemonics -> ECDSA key -> ECDSA key doesn't yield same public key

### ISSUE THREAD

**Issue by gregscullard** [hashgraph/hedera-sdk-js#1396]
### Description

I'm recovering a private key from Mnemonics and printing the public key as a string.
I then use the private key string above to instantiate a PrivateKey and print its public key as a string.
The two public keys differ, yet printing both private keys as string yield the same result, see example below

### Steps to reproduce

```
import {
    Mnemonic, PrivateKey,
} from "@hashgraph/sdk";

async function test() {
    const mnemonic = await Mnemonic.fromString("hamster produce dry base sunny bubble disease throw cricket garden beyond script");
    const privateKey = await mnemonic.toEcdsaPrivateKey();

    const privateKeyString = privateKey.toStringRaw();
    const publicKeyString = privateKey.publicKey.toStringRaw();

    // Restore from string
    const restoredPrivateKey = PrivateKey.fromStringECDSA(privateKeyString);
    const restoredPrivateKeyString = privateKey.toStringRaw();
    const restoredPublicKeyString = restoredPrivateKey.publicKey.toStringRaw();

    const results = {
      FromMnemonics: {Private: privateKeyString, Public: publicKeyString},
      ECDSA: {Private: restoredPrivateKeyString, Public: restoredPublicKeyString}
    };
    console.table(results);

}

test();
```

outputs

```
┌───────────────┬────────────────────────────────────────────────────────────────────┬──────────────────────────────────────────────────────────────────────┐
│    (index)    │                              Private                               │                                Public                                │
├───────────────┼────────────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────┤
│ FromMnemonics │ '63167e08a41aaaa8226945b999a201f789e1a362d4d5b93d605e0841795eb500' │ '0322c1cc8bdbdd54b851251a61ea98d4da31f2749f00e44d66efc2295df76b4a9a' │
│     ECDSA     │ '63167e08a41aaaa8226945b999a201f789e1a362d4d5b93d605e0841795eb500' │ '0283ddc3dfca5dab9c10b4a6be831bae1b50ab36224c305bdd008c6042b34b3123' │
└───────────────┴────────────────────────────────────────────────────────────────────┴──────────────────────────────────────────────────────────────────────┘
```

Shouldn't both public keys be the same ?

Note, investigating through debugging the SDK shows that the two private keys differ slightly, see below

First key (from mnemonics - chaincode attribute value removed)
```
{ "_key": { "_keyPair": { "privateKey": { "type": "Buffer", "data": [ 99, 22,
126, 8, 164, 26, 170, 168, 34, 105, 69, 185, 153, 162, 1, 247, 137, 225, 163,
98, 212, 213, 185, 61, 96, 94, 8, 65, 121, 94, 181 ] }, 
"publicKey": { "type":
"Buffer", "data": [ 3, 34, 193, 204, 139, 219, 221, 84, 184, 81, 37, 26, 97,
234, 152, 212, 218, 49, 242, 116, 159, 0, 228, 77, 102, 239, 194, 41, 93, 247,
107, 74, 154 ] } }
  }
```

Second key (fromStringECDSA)
```
Second _key { "_keyPair": { "privateKey": { "type": "Buffer", "data": [ 99, 22,
126, 8, 164, 26, 170, 168, 34, 105, 69, 185, 153, 162, 1, 247, 137, 225, 163,
98, 212, 213, 185, 61, 96, 94, 8, 65, 121, 94, 181, 0 ] },
"publicKey": {
"type": "Buffer", "data": [ 2, 131, 221, 195, 223, 202, 93, 171, 156, 16, 180,
166, 190, 131, 27, 174, 27, 80, 171, 54, 34, 76, 48, 91, 221, 0, 140, 96, 66,
179, 75, 49, 35 ] } }, "_chainCode": null
}
```

note that the first contains an extra element at the end of the buffer `0`, the public key buffer is vastly different.

### Hedera network

mainnet, testnet, previewnet

### Version

v2.19.1 and 2.19.0

### Operating system

None

**Comment by gregscullard** [comment]
May be related to https://github.com/hashgraph/hedera-sdk-js/issues/1350
Note: Any changes should carefully consider the impact on existing users/accounts. 
Accounts created using a public key from mnemonics will not work if they are later recovered and the SDK generates a different public key as a result of the change.
Likewise, accounts created using a public fromString will not work if they are later recovered with the same private key string and the SDK generates a different public key as a result of the change.

**Comment by ochikov** [comment]
Hello @gregscullard and thank you for your issue. We are currently discussing this in the [reference issue](https://github.com/hashgraph/hedera-sdk-reference/issues/73). 
We plan to work as soon as we clear every questions around the refactor.

**Comment by alittley** [comment]
I don't think this problem is related to the derivation path issue.

@gregscullard Looking at the raw bytes you posted, I'm only counting `31` bytes in the first private key. You say that is what came out of `mnemonic.toEcdsaPrivateKey()`? Private keys ought to be `32` bytes.
```
[ 99, 22, 126, 8, 164, 26, 170, 168, 34, 105, 69, 185, 153, 162, 1, 247, 137, 225, 163, 98, 212, 213, 185, 61, 96, 94, 8, 65, 121, 94, 181 ]
```

This also explains why the next key has an "extra" `0`. `PrivateKey.fromStringECDSA` is probably just blindly reading the values into a properly sized `32` byte array.

Next step is to get confirmation of whether `mnemonic.toEcdsaPrivateKey()`, *always* yields a `31` byte key, or if this is intermittent.

**Comment by ochikov** [comment]
We are going to investigate the cryptography package and return with a potential fix.

**Comment by petreze** [comment]
@gregscullard @alittley 
Fixed with the following [PR](https://github.com/hashgraph/hedera-sdk-js/pull/1397) and added an unit test as well. It will be available after the next release of the cryptography package

**Comment by alittley** [comment]
I think it is very important that we figure out *why* that particular derivation path is returning a private key of 31 bytes.

Something is very wrong here: the existing (non-hardened) derivation path *ought* to yield a valid key. `[44, 3030, 0, 0]` isn't the derivation path we want to be using, but it is still a *valid* derivation path.

I think merging a fix for this should be delayed for 2 reasons:
1. We need a full understanding of what is going wrong with the current key derivation. How can it be that a certain derivation path yields non-sensical keys?
2. IMO even once this has been clarified, the fix should be delayed until the resolution of the mentioned epic. There should be one more change to how keys are handled, and we should be certain that the new implementation is correct.

**Comment by ochikov** [comment]
@alittley That's why we put the PR in draft. We are on the same page that when we clarify the reason, we need to include the fix in the Epic for the Mnemonic refactor. @petreze is still researching it.

**Comment by petreze** [comment]
@alittley 
As I mentioned, whenever we are on the last iteration of indexes of the wrong derivation path [`hex.decode(secp256k1.keyFromPrivate(ki.toArray()).getPrivate("hex"))`](https://github.com/hashgraph/hedera-sdk-js/blob/develop/packages/cryptography/src/primitive/bip32.js#L65-L67) returns a ByteArray with 31 length. Since after each iteration, we are overwriting the value of `keyData`, this last iteration over the second `0` index strongly depends on the previous outcomes when [bip32.derive()](https://github.com/hashgraph/hedera-sdk-js/blob/develop/packages/cryptography/src/Mnemonic.js#L207) is called with the other (44, 3030, 0) indexes of the derivation path. This means that by changing the indexes from not hardened to hardened indexes, so from `[44, 3030, 0, 0] ` to `[44 | 0x80000000, 3030 | 0x80000000, 0 | 0x80000000, 0]`, every private key derived from the previous one will be different if we change the path.

Meaning that, if using the right derivation path, it should generate the correct private key and it does so. We are using `elliptic` JS library with the `secp256k1` elliptic curve and the mentioned line `secp256k1.keyFromPrivate(ki.toArray()).getPrivate("hex")` comes from there...

Furthermore, we have an unit [test](https://github.com/hashgraph/hedera-sdk-js/blob/develop/packages/cryptography/test/unit/Mnemonic.js#L61-L92) inside the cryptography package which compares all of the derived keys, generated from a specific mnemonic, with the predefined correct ones for that same mnemonic phrase described [here](https://gist.github.com/poma/6b503eead48628a03088fd270a68e369)

So it is safe to say that this should be the right approach

**Comment by alittley** [comment]
Like I mentioned above, `[44, 3030, 0, 0]` is one of infinitely many arbitrary, but valid, derivation paths. Even though no one is ever likely to choose it, it should still be possible to do so. No derivation path, even the "wrong" one currently in use, should ever yield a key of 31 bytes.

The fact that we somehow have a derivation path which *does* yield an invalid key speaks to there being a bug somewhere, whether in the external library, or its usage. Unless there is something fundamental I'm missing here

**Comment by gregscullard** [comment]
The oddity in the test sample I provided is that it's a particular set of mnemonics that don't work, mnemonics generated from the current SDK do work. 
The unit test may indeed pass, but not with these mnemonics (hamster produce dry base sunny bubble disease throw cricket garden beyond script).

**Comment by alittley** [comment]
I don't think the linked PR does anything to address the root cause of this issue. Like Greg said, there are weird inconsistencies still happening. This isn't changed by the fact that using a different derivation path makes the issue disappear for a couple test mnemonic phrases.

I would be curious if a test that generates a million different mnemonic phrases and tests their derivation would reveal any other instances of 31 byte keys.

**Comment by petreze** [comment]
@alittley @gregscullard 
I've tested a random mnemonic generation and the behaviour of the SDK
So, whenever we generate a mnemonic with `Mnemonic.generate12()` (this takes 12 words from the official [BIP39 list](https://getcoinplate.com/blog/official-bip39-word-list-mnemonic-in-english-verified/) of words that generate a proper entropy and checksum), there is a chance of getting a 31 bytes key. From the tests, I can conclude that it does not matter if we use the right derivation path `[44 | HARDEDED, 3030 | HARDEDED, 0 | HARDEDED, 0]` or the wrong one `[44, 3030, 0, 0]`, it can still generate at random times 31 bytes keys

But the randomness does not stop here... As you remember, I mentioned that when we call `Mnemonic.toEcdsaPrivateKey()` we iterate over the indexes of the derivation path and on each iteration, we overwrite the previous private key value with the newly derived key data.
So lets say we are using `[44, 3030, 0, 0]` derivation path (for simplicity).

- the first iteration over index `44` yields `pk1` which is the returned `keyData` value
- the second one iterates over index `3030` with `pk1` as `parentKey` and it returns the derived key `pk2` from its parent key `pk1`
- this process continues until we get to the last index and the last returned value is in fact the private key at this specified path

The strange behaviour where we can get a 31 bytes private key could happen on each of these iterations randomly for a random mnemonic phrases. It just happened that in Greg's case the last iteration returns 31 bytes key and we found the bug but up until now I could not find any relations between the generated mnemonics and the cases where we yield 31 bytes

**Comment by alittley** [comment]
@petreze @ochikov It sounds to me like the bug is in `bip32::derive`.

In particular, I'm suspicious of [this return statement](https://github.com/hashgraph/hedera-sdk-js/blob/021459bdafae6d529d4687b14cce9a2845033241/packages/cryptography/src/primitive/bip32.js#L64)
```
        return {
            keyData: hex.decode(         // << suspicious
                secp256k1.keyFromPrivate(ki.toArray()).getPrivate("hex") 
            ),
            chainCode: IR,
        };
```

`hex.decode` must be returning an array *without 0 padding* in the case that the most significant byte of the resulting key is `0x0`. I believe this case should arise 1/16 times (ouch).

**Comment by petreze** [comment]
@alittley 
Will investigate that, but yes this part is a bit uncertain and one of my suggestions was also that statement, since the other part of the `derive` functionality LGTM

**Comment by petreze** [comment]
@alittley 
Hello Austin, I think the issue might be in fact _`returning an array without 0 padding in the case that the most significant byte of the resulting key is 0x00`_

I dug into `ethers.js` implementation of how they handle the situation.
They are using [bytes32](https://github.com/ethers-io/ethers.js/blob/bf0b468490cb293cd916e4fff06e0909273719e6/packages/hdnode/src.ts/index.ts#L203) function in order to generate the private key on each index iteration. Under the hood it uses [hexZeroPad](https://github.com/ethers-io/ethers.js/blob/bf0b468490cb293cd916e4fff06e0909273719e6/packages/bytes/src.ts/index.ts#L315) which I think handles the 0 padding case

In our implementation, as I mentioned `elliptic` JS library is used for the whole generation and derivation of the private keys and there is an option for the `toArray()` method to specify the endian and zero pad to length if already exceeding
Basically the equivalent of their implementation. When I set the length to 32, it always generates 32 bytes private keys.. but the [test](https://github.com/hashgraph/hedera-sdk-js/blob/develop/packages/cryptography/test/unit/Mnemonic.js#L61-L92) with the predefined mnemonic to set of private key derivation fails. This is defined [here](https://gist.github.com/poma/6b503eead48628a03088fd270a68e369) and it was the only test for correctly generated and derived ecdsa private keys in our SDK

**Comment by petreze** [comment]
@alittley 
Short update
The functionality from the library did not work, so I've implemented the zero padding in the SDK itself. I also added an ecdsa generation from a specific mnemonic with the default ethereum derivation path `44'/60'/0'/0/0` which is used to test in **ethers.js** as well. Everything is pushed in the same [branch](https://github.com/hashgraph/hedera-sdk-js/pull/1397) so you can check it

**Comment by alittley** [comment]
@ochikov @petreze I'm not sure if it has been discussed yet, but the fix here requires separate consideration from the [derivation path bug](https://github.com/hashgraph/hedera-sdk-js/issues/1350).

Whereas a "compatibility mode" for keys generated at the incorrect derivation path can be achieved by manually specifying the old path, that won't be possible for keys that fall victim to this derivation bug. The existing bad bip32 derivation implementation will have to be preserved in some way, to rescue the ~1.6% of people who are affected by this additional bug.

**Comment by petreze** [comment]
Fixed with [PR](https://github.com/hashgraph/hedera-sdk-js/pull/1550)
